File: java/src/test/java/shapes/CommonsTest.java
package shapes;

import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import shapes.base.commons.Dimensional;
import shapes.fixtures.Helper;

@DisplayName("Commons Tests")
class CommonsTest {
    @Nested
    @DisplayName("Dimensional Tests")
    class DimensionalTest {
        @Test
        @DisplayName("Test Circle Implements Dimensional")
        void testCircleImplementsDimensional() {
            var circle = Helper.setupCircle();
            assertDimensional(circle, 754.767, 97.389);
        }

        @Test
        @DisplayName("Test Rectangle Implements Dimensional")
        void testRectangleImplementsDimensional() {
            var rectangle = Helper.setupRectangle();
            assertDimensional(rectangle, 756.437, 112.0);
        }

        @Test
        @DisplayName("Test Square Implements Dimensional")
        void testSquareImplementsDimensional() {
            var square = Helper.setupSquare();
            assertDimensional(square, 105.062, 41.0);
        }
    }

    void assertDimensional(
        Dimensional dimensionalShape,
        double expectedArea,
        double expectedPerimeter
    ) {
        double areaDiff = Math.abs(dimensionalShape.area() - expectedArea);
        double perimeterDiff = Math.abs(dimensionalShape.perimeter() - expectedPerimeter);

        assertTrue(areaDiff < 0.001, "AREA NOT EQUAL");
        assertTrue(perimeterDiff < 0.001, "PERIMETER NOT EQUAL!");
    }
}

File: elixir/test/shapes/commons_test.exs
alias Fixtures.Helper
alias Shapes.Base.Commons.Dimensional

defmodule TestDimensional do
  use ExUnit.Case, async: true

  describe "Dimensional" do
    test "circle implements dimensional" do
      circle = Helper.setup_circle()
      assert_dimensional(circle, 754.767, 97.389)
    end

    test "rectangle implements dimensional" do
      rectangle = Helper.setup_rectangle()
      assert_dimensional(rectangle, 756.437, 112.0)
    end

    test "square implements dimensional" do
      square = Helper.setup_square()
      assert_dimensional(square, 105.062, 41.0)
    end
  end

  defp assert_dimensional(
         dimensional_shape,
         expected_area,
         expected_perimeter
       ) do
    area_diff = abs(Dimensional.area(dimensional_shape) - expected_area)
    perimeter_diff = abs(Dimensional.perimeter(dimensional_shape) - expected_perimeter)

    assert area_diff < 0.001, "AREA NOT EQUAL!"
    assert perimeter_diff < 0.001, "PERIMETER NOT EQUAL!"
  end
end
